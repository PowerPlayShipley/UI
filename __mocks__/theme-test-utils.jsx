// TODO: Fix this??

import * as React from 'react'
import userEvent from '@testing-library/user-event'
import { render as rtlRender } from '@testing-library/react'

import { ThemeProvider } from "styled-components";

import theme from '../src/themes'
import GlobalStyle from "../src/components/GlobalStyle";

function renderWithTheme(ui, { ...options } = {}) {
  const Wrapper = ({ children }) => (
    <ThemeProvider theme={theme}>
      <GlobalStyle />
      {children}
      <div id='modal-root' />
    </ThemeProvider>
  )

  return rtlRender(ui, { wrapper: Wrapper, ...options })
}

// THE BELLOW COME FROM `https://github.com/styled-components/styled-components/blob/0aa3170c255a49cd41c3fbeb2b8051b5d132f229/src/test/utils.js`

const stripComments = (str) => str.replace(/\/\*.*?\*\/\n?/g, '');

const stripWhitespace = (str) => str
  .trim()
  .replace(/([;\{\}])/g, '$1  ')
  .replace(/\s+/g, ' ');

const getCSS = (scope) =>
  Array.from(scope.querySelectorAll('style'))
    .map(tag => tag.innerHTML)
    .join('\n')
    .replace(/ {/g, '{')
    .replace(/:\s+/g, ':')
    .replace(/:\s+;/g, ':;');

function expectCSSMatches ( _expectation, opts = { ignoreWhitespace: true }) {
  // NOTE: This should normalise both CSS strings to make irrelevant mismatches less likely
  const expectation = _expectation
    .replace(/ {/g, '{')
    .replace(/:\s+/g, ':')
    .replace(/:\s+;/g, ':;');

  const css = getCSS(document);

  if (opts.ignoreWhitespace) {
    const stripped = stripWhitespace(stripComments(css));
    expect(stripped).toEqual(stripWhitespace(expectation));
    return stripped;
  } else {
    expect(css).toEqual(expectation);
    return css;
  }
};

export * from '@testing-library/react';

export { renderWithTheme, expectCSSMatches };
