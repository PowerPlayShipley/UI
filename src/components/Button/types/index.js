export { default as primary } from './primary'
export { default as outline } from './outline'
export { default as danger } from './danger'
