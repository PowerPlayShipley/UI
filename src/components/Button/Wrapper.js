import styled, { css } from "styled-components";
import { darken } from 'polished'

import { hexToRGB } from "../../themes/utils";

import * as types from './types'

const Wrapper = styled.button`
  position: relative;
  display: inline-block;
  
  padding: 5px ${({ theme }) => theme.main.sizes.spacing['space-3']};
  
  font-size: ${({ theme }) => theme.main.typography["font-size-small"]};
  font-weight: ${({ theme }) => theme.main.typography.fontWeights.bold};
  line-height: 20px;
  
  white-space: nowrap;
  vertical-align: middle;
  
  cursor: pointer;
  user-select: none;
  border: ${({ theme }) => `${theme.main.misc.borderWidth} ${theme.main.misc.borderStyle}`};
  border-radius: ${({ theme }) => theme.main.misc.radius.default };
  
  appearance: none;
  
  &:hover {
    text-decoration: none;
  }

  &:disabled, &.disabled, &[aria-disabled=true] {
    cursor: default;
  }

  .dropdown-caret {
    margin-left: ${({ theme }) => theme.main.sizes.spacing['space-1']};
    opacity: 0.8;
  }
  
  & .btn-icon {
    margin-right: ${({theme}) => theme.main.sizes.spacing['space-1']};
    color: inherit;
    vertical-align: text-bottom;

    &:only-child {
      margin-right: 0;
    }
  }
  
  // Default button tags
  & {
    color: ${({ theme }) => theme.main.colors.text.default };
    background-color: ${({ theme }) => theme.main.colors.backgrounds.default };
    border: ${({ theme }) => theme.main.misc.border };
    
    box-shadow: ${({ theme }) => theme.main.misc.shadow.default }, ${({ theme }) => theme.main.misc.inset.default };
    
    transition: 0.2s cubic-bezier(0.3, 0, 0.5, 1);
    transition-property: color, background-color, border-color;

    &:hover, &.hover, [open] > & {
      background-color: ${({ theme }) => theme.main.colors.system.grays['100'] };
      transition-duration: 0.1s;
    }

    &:active {
      background-color: ${({ theme }) => darken('0.03', theme.main.colors.system.grays['100']) };
      transition: none;
    }

    &.selected, &[aria-selected=true] {
      background-color: ${({ theme }) => darken('0.02', theme.main.colors.system.grays['100']) };
      box-shadow: ${({ theme }) => theme.main.misc.inset.medium};
    }

    &:disabled, &.disabled, &[aria-disabled=true] {
      color: ${({ theme }) => theme.main.colors.text.disabled };
      background-color: ${({ theme }) => theme.main.colors.backgrounds.default };
      cursor: default;

      .btn-icon {
        color: ${({ theme }) => theme.main.colors.text.disabled };
      }
    }

    // Keep :focus after :disabled. Allows to see the focus ring even on disabled buttons
    &:focus, &.focus {
      outline: none;
      box-shadow: 0 0 0 3px rgba(${({ theme }) => hexToRGB(theme.main.colors.system.blues['500'])}, 0.15);
    }
  }
  
  ${({ state }) => types[state]}
`

export default Wrapper
