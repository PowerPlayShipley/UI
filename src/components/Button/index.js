/**
 * Simple buttons, styled similar to github's primer css.
 * */

import React, { cloneElement } from 'react'
import PropTypes from 'prop-types'

import clsx from "clsx";

import Wrapper from "./Wrapper";

const Button = ({ as, disabled, state, title, className, children, icon, ...rest }) => {
  return (
    <Wrapper className={clsx(className, state && state)} as={as} disabled={disabled} state={state} {...rest}>
      {(icon && !children) && cloneElement(icon, { className: 'btn-icon' })}
      {title && title || children}
    </Wrapper>
  )
}

Button.defaultProps = {
  as: 'button',
  disabled: false,
  state: 'default'
}

Button.propTypes = {
  as: PropTypes.oneOf(['div', 'a', 'button']),
  children: PropTypes.element,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  state: PropTypes.oneOf(['default', 'primary', 'outline', 'danger']),
  title: PropTypes.string,
  onClick: PropTypes.func,
  icon: PropTypes.element
}

export default Button
