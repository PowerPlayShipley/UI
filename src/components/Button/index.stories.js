import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faPlus } from "@fortawesome/free-solid-svg-icons"

import Button from "./index";

export default {
  title: 'Button',
  component: Button
}

const Template = (args) => (<Button {...args} />);

export const Default = Template.bind({});
Default.args = {
  title: 'Default',
  state: 'default',
  disabled: false,
  as: 'button'
}

export const Primary = Template.bind({});
Primary.args = {
  title: 'Primary',
  state: 'primary',
  disabled: false,
  as: 'button'
}

export const Outline = Template.bind({});
Outline.args = {
  title: 'Outline',
  state: 'outline',
  disabled: false,
  as: 'button'
}

export const Danger = Template.bind({});
Danger.args = {
  title: 'Danger',
  state: 'danger',
  as: 'button'
}

export const Disabled = Template.bind({});
Disabled.args = {
  title: 'Disabled',
  state: 'default',
  disabled: true,
  as: 'button'
}

export const WithIcon = Template.bind({});
WithIcon.args = {
  state: 'primary',
  icon: <FontAwesomeIcon style={{ marginRight: '4px' }} icon={faPlus} />,
  title: " New",
  disabled: false,
  as: 'a'
}

export const UsingChildren = (args) => (
  <Button {...args}>
    Children
  </Button>
)
UsingChildren.args = {
  state: 'primary',
  disabled: false,
  as: 'button'
}
