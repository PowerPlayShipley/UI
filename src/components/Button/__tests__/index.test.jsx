/**
 * @jest-environment jsdom
 */

import React from 'react'
import { renderWithTheme, fireEvent } from '../../../../__mocks__/theme-test-utils'

import Button from "../index";

describe('Display Button', () => {
  it('should render cell', () => {
    const { container } = renderWithTheme(<Button title={'Hello'} />)
    expect(container).toHaveTextContent('Hello')
  })

  it('should create valid snapshot', function () {
    const { container } = renderWithTheme(<Button title={'Hello'} />)
    expect(container.firstChild).toMatchSnapshot()
  });

  it('should fire button', function () {
    const handleClick = jest.fn()

    const { container } = renderWithTheme(<Button title={'Hello'} onClick={handleClick} />)

    fireEvent.click(container.querySelector('button'))
    expect(handleClick).toHaveBeenCalledTimes(1)
  });

  it(`should create valid button state - default`, function () {
    const { container } = renderWithTheme(<Button title={'Hello'} state={'default'} />)
    expect(container.firstChild).toHaveClass('default')
  });

  it(`should create valid button state - primary`, function () {
    const { container } = renderWithTheme(<Button title={'Hello'} state={'primary'} />)
    expect(container.firstChild).toHaveClass('primary')
  });

  it(`should create valid button state - outline`, function () {
    const { container } = renderWithTheme(<Button title={'Hello'} state={'outline'} />)
    expect(container.firstChild).toHaveClass('outline')
  });

  it(`should create valid button state - danger`, function () {
    const { container } = renderWithTheme(<Button title={'Hello'} state={'danger'} />)
    expect(container.firstChild).toHaveClass('danger')
  });

  it(`should create valid button - div`, function () {
    const { container } = renderWithTheme(<Button title={'Hello'} as={'div'} />)
    expect(container.firstChild.nodeName).toBe('DIV')
  });

  it(`should create valid button - link`, function () {
    const { container } = renderWithTheme(<Button title={'Hello'} as={'a'} />)
    expect(container.firstChild.nodeName).toBe('A')
  });

  it(`should create valid button - button`, function () {
    const { container } = renderWithTheme(<Button title={'Hello'} as={'button'} />)
    expect(container.firstChild.nodeName).toBe('BUTTON')
  });
})
