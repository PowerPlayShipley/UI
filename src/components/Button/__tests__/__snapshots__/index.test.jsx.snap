// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Display Button should create valid snapshot 1`] = `
.c0 {
  position: relative;
  display: inline-block;
  padding: 5px 16px;
  font-size: 14px;
  font-weight: 600;
  line-height: 20px;
  white-space: nowrap;
  vertical-align: middle;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border: 1px solid;
  border-radius: 6px;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c0:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:disabled,
.c0.disabled,
.c0[aria-disabled=true] {
  cursor: default;
}

.c0 .dropdown-caret {
  margin-left: 4px;
  opacity: 0.8;
}

.c0 .btn-icon {
  margin-right: 4px;
  color: inherit;
  vertical-align: text-bottom;
}

.c0 .btn-icon:only-child {
  margin-right: 0;
}

.c0 {
  color: #24292f;
  background-color: #f6f8fa;
  border: 1px rgba(208,215,222,0.5) solid;
  box-shadow: 0 1px 0 rgba(27,31,36,0.04),inset 0 1px 0 rgba(255,255,255,0.25);
  -webkit-transition: 0.2s cubic-bezier(0.3,0,0.5,1);
  transition: 0.2s cubic-bezier(0.3,0,0.5,1);
  -webkit-transition-property: color,background-color,border-color;
  transition-property: color,background-color,border-color;
}

.c0:hover,
.c0.hover,
[open] > .c0 {
  background-color: #eaeef2;
  -webkit-transition-duration: 0.1s;
  transition-duration: 0.1s;
}

.c0:active {
  background-color: #e1e6ec;
  -webkit-transition: none;
  transition: none;
}

.c0.selected,
.c0[aria-selected=true] {
  background-color: #e4e9ee;
  box-shadow: inset 0 1px 0 rgba(208,215,222,0.2);
}

.c0:disabled,
.c0.disabled,
.c0[aria-disabled=true] {
  color: #8c959f;
  background-color: #f6f8fa;
  cursor: default;
}

.c0:disabled .btn-icon,
.c0.disabled .btn-icon,
.c0[aria-disabled=true] .btn-icon {
  color: #8c959f;
}

.c0:focus,
.c0.focus {
  outline: none;
  box-shadow: 0 0 0 3px rgba(9,105,218,0.15);
}

<button
  class="c0 default"
>
  Hello
</button>
`;
